cversion/tdu.1
.\" Process this file with
.\" groff -man -Tascii tdu.1
.\"
.\" tdu - a text-mode disk usage visualization utility
.\" Copyright (C) 2004 Darren Stuart Embry.  
.\" 
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\" 
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307$
.\"
.TH tdu 1
.SH NAME
tdu \- folding-outline disk usage display
.SH SYNOPSIS
.B "du"
.IR "ARG" " ... |"
.B "tdu"
.RI "[" "OPTION" " ...]"
.SH DESCRIPTION
tdu is a program that displays disk space utilization as reported by du(1) in
an interactive full-screen folding outline.  tdu displays the space taken by
each directory, including all its descendents.  The user can navigate the tree
structure and open (or close) directories in order to display (or hide) the
utilization of subdirectories (and files, if du -x is used) within each
directory.  The entries can be sorted by filename or space utilizied, or
reverted back to their original sorting order.
.SH KEYS
.SS Navigation
.IP "UP, DOWN, PAGEUP, PAGEDOWN"
Move the cursor up or down by one line or page.
.IP "HOME or <"
Move to cursor to the first line.
.IP "END or >"
Move to cursor to the last line.
.IP "p, P"
Move the cursor to the current item's parent directory.
.IP "Control-L"
Move the cursor line to the center of the screen.
.SS Emacs(1) and vi(1) style keybindings
.IP "Control-P, Control-N, Control-B, Control-F"
emacs(1) style keybindings for the arrow keys.
.IP "h, j, k, l"
vi(1) style keybindings for the arrow keys.
.SS Expanding and Collapsing
.IP "LEFT or 0 (zero)"
Collapse the current node's children.
.IP "RIGHT"
Expand the current node's chlidren.
Repeated presses of this key expand further.
.IP "1 to 9"
Expand 1 to 9 levels deep.
.IP "*"
Expand all levels deep.
.SS Sorting
.IP "s, S"
Sort current item's children in ascending, descending order by size.
.IP "n, N"
Sort current item's children in ascending, descending order by name.
.IP "d, D"
Sort current item's children in ascending, descending order by
number of descendent nodes.
.IP "u, U"
Revert current item's children to original order, or its reverse.
.IP "="
Press before one of the above sorting keys to apply the sort recursively.
.SS Miscellaneous
.IP "#"
Hide/show the number of descendents of each node.
.IP "a, A"
Toggle the use of ASCII line-drawing characters.
.IP "Control-R"
Refresh the display.
.IP "q, x, ESC"
Quit the program.
.IP "?"
Display online help.
.IP "c, C"
Display copyright and licensing information.
.SH OPTIONS
.IP "-h, --help"
Display a message listing the command-line invocation synopsis and options.
.IP "-A, --ascii-tree"
Use ASCII characters instead of actual line-drawing characters.
Interactively, This is also toggled using the "a" or "A" key.
.IP "-V, --show-version"
Display the version and copyright information.
.IP "-I, -G"
These options have been removed from tdu.
You can use a separate program, dugroup(1), to group files together by
extension.
.SH BUGS AND LIMITATIONS
The use of certain options for du will or may cause this program to
not work properly.  Most if not all of these limitations also affect
xdu(1).

This program cannot handle du's -h/--human-readable option, since it
expects to read integer block sizes.

du's -c/--total option will mislead tdu.

du's -b/--bytes option may produce integers too large for tdu to
store, depending on the platform on which tdu was compiled.

This program does not currently have a -0/--null option to handle du's
output if du is supplied with the same option.

du's -S/--separate-dirs option may mislead tdu.
.SH AUTHOR
Darren Stuart Embry (dse@webonastick.com).
.SH COPYRIGHT
This program is free software.  It can be distributed and/or modified under
the terms of the GNU General Public License, Version 2.  For more details you
can view it at http://webonastick.com/tdu/COPYING.txt
.SH "SEE ALSO"
.BR du (1),
.BR xdu (1),
.BR dugroup (1).


cversion/tdu.c
/*
 * tdu.c
 * This file is part of tdu, a text-mode disk usage visualization utility.
 *
 * Copyright (C) 2004-2008 Darren Stuart Embry.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *                                                                             
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *                                                                             
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307$
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>

#include "tdu.h"
#include "node.h"
#include "tduint.h"

static char *optstring = "hG:I:AVP";
static char *progname = "tdu";

struct option long_options[] = {
	{ "help",       0, NULL, 'h' },
	{ "ascii-tree", 0, NULL, 'A' },
	{ "parse-only", 0, NULL, 'P' },
	{ "version",    0, NULL, 'V' },
	{ "debug",      0, NULL, 'd' },
	{ NULL,         0, NULL, 0 }
};

#define TDU_USAGE_MESSAGE \
	"usage: du [OPTION ...] [FILE ...] | %s [OPTION ...] [FILE ...]\n" \
	"  -h, --help        display this message\n" \
	"  -A, --ascii-tree  display tree branches using ASCII characters\n" \
	"  -V, --version     show version, license terms\n"

void
version_exit (int status)
{
	fprintf(stderr, TDU_COPYRIGHT_INFO);
	exit(status);
}

void
usage_exit (int status)
{
	fprintf(stderr, TDU_USAGE_MESSAGE, progname);
	exit(status);
}

typedef struct options {
	bool help;
	int optind;
	bool parse_only;
} options_s;

options_s *
get_options (int argc, char **argv)
{
	options_s *options;
	int c;

	options = (options_s *)malloc(sizeof(options_s));
	if (options == NULL) {
		perror("get_options: malloc");
		exit(1);
	}
	options->help = 1;
	options->optind = -1;
	options->parse_only = 0;

	while ((c = getopt_long(argc, argv, optstring,
				long_options, NULL)) != -1) {
		switch (c) {
		case 'G':
			fprintf(stderr,
				"-G option is no longer implemented in tdu.\n"
				"Its capability is part of dugroup, "
				"a separate program.\n");
			exit(1);
		case 'I':
			fprintf(stderr,
				"-I option is no longer implemented in tdu.\n"
				"Its capability is part of dugroup, "
				"a separate program.\n");
			exit(1);
		case 'A':
			ascii_tree_chars = 1;
			break;
		case 'h':
			usage_exit(0);
			break;
		case 'V':
			version_exit(0);
			break;
		case 'P':
			options->parse_only = 1;
			break;
		default:
			usage_exit(1);
			break;
		}
	}

	options->optind = optind;
	return options;
}

int
main (int argc, char **argv)
{
	node_s *node;
	options_s *options;

	if (NULL == (options = get_options(argc, argv))) {
		--argc, ++argv;
	} 
	else {
		argc -= options->optind;
		argv += options->optind;
	}

	node = parse_file(*argv);

	if (options->parse_only) {
		return 0;
	}

	if (node) {
		expand_tree(node, 1);
		tdu_interface_run(node);
	}

	return 0;
}

int
debug (const char *fmt, ...)
{
	va_list ap;
	va_start(ap, fmt);
	va_end(ap);

	if (getenv("DEBUG") && !isatty(fileno(stderr))) {
		int result = vfprintf(stderr, fmt, ap);
		fflush(stderr);
		return result;
	}
	else {
		return 0;
	}
}
